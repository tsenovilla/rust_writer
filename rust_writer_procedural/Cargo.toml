[package]
name = "rust_writer_procedural"
version.workspace = true
repository.workspace = true 
license.workspace = true
rust-version.workspace = true
edition.workspace = true
description.workspace = true
documentation.workspace = true

[dependencies]
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true, features = ["full", "extra-traits", "parsing"] }

[lib]
proc-macro = true

[dev-dependencies]
trybuild = "1.0.103"
rust_writer = { workspace = true }
test_builder = { workspace = true }
