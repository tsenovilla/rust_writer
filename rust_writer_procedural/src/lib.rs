// SPDX-License-Identifier: GPL-3.0

mod finder;
pub(crate) mod helpers;
mod mutator;
pub(crate) mod parse;

use proc_macro::TokenStream;

/// # Description
///
/// The mutator macro is used to define a new implementor which combines a  
///
/// # Compatibility with [`#[finder]`](https://docs.rs/rust_writer/latest/rust_writer/ast/attr.finder.html)
#[proc_macro_attribute]
pub fn mutator(attrs: TokenStream, item: TokenStream) -> TokenStream {
	mutator::mutator(attrs, item)
}

#[proc_macro_attribute]
pub fn local_mutator(_: TokenStream, item: TokenStream) -> TokenStream {
	mutator::local_mutator(item)
}

#[proc_macro_attribute]
pub fn finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::finder(attrs, item)
}

#[proc_macro_attribute]
pub fn local_finder(attrs: TokenStream, item: TokenStream) -> TokenStream {
	finder::local_finder(attrs, item)
}

#[doc(hidden)]
#[proc_macro_attribute]
pub fn already_expanded(_: TokenStream, item: TokenStream) -> TokenStream {
	item
}
