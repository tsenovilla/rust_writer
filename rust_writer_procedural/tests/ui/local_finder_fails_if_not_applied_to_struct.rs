// SPDX-License-Identifier: GPL-3.0

use rust_writer_procedural::local_finder;

#[local_finder('a)]
enum SomeStruct {
	A,
	B,
	C,
}

fn main() {}
