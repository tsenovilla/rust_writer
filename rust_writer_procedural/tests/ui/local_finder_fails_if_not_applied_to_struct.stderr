error: expected `struct`
 --> tests/ui/local_finder_fails_if_not_applied_to_struct.rs:6:1
  |
6 | enum SomeStruct {
  | ^^^^
